### ISSUE TITLE
How does throw_on_cancel work in tp.system.prompt()

### ISSUE THREAD

**Issue by Deeds87** [SilentVoid13/Templater#1294]
**Plugin information:**

-   OS: [Windows 11]
-   Templater version: [2.1.0]
-   Obsidian version: [1.5.3]

**Describe the bug**
When i use
`let topic = await tp.system.prompt("Set topic", "this is a default value", true, false);`
there is no error in the console and no error modal shown up when i hit "esc".
The file will be created as Untitled.md.

**Expected behavior**
I am expecting the same behaviour as from the suggester comand.
There i get a modal with the cancel error and the file will not be created.

Perhaps i missunterstood how to use throw_on_cancel and there is no bug. If so please help me to unterstand how to achive the desired behaviour.

Thanks for this great plugin

**Comment by Zachatoo** [comment]
I can't reproduce. Here's the script I'm testing with.

```js
<%*
let topic = await tp.system.prompt("Set topic", "this is a default value", true, false);
console.log('I will not be hit due to error');
%>
```

If I execute this script and hit escape or click the X button in the modal, the `console.log` will not be hit because an exception will be thrown.

If I hit enter, then the `console.log` will be hit.

Video below:

https://github.com/SilentVoid13/Templater/assets/6936914/99c424d3-3967-45ad-8ce9-86e6b661e9f4

**Comment by Deeds87** [comment]
I couldn't find my mistake but you are totally right, after trying it again it works like you say.
But what would be more important to me is whether there is a way to interrupt the file creation? Like with suggester.

**Comment by Zachatoo** [comment]
That depends on how the file is being created. There's no way to interrupt the file being created if you're using the `Templater: Create new note from template` command, or if you created the file some other way and Templater is executed because of the `Trigger Templater on new file creation` setting. In both of these cases, the file is created _before_ your template starts executing, it's already too late. The best you could do is delete the file after the fact.

```js
<%*
try {
  const topic = await tp.system.prompt("Set topic", "this is a default value", true, false);
  // other stuff
} catch {
  // if prompt is cancelled (or other error happens), delete file that was created
  const file = tp.file.find_tfile(tp.file.path(true));
  await app.vault.delete(file);
}
-%>
```

You may want to instead consider a two template approach, using the `Templater: Open Insert Template Model` command. This command will not create a new file, it will insert the template into the current file and execute it immediately.

The first template will do any prompts that you need in order to create the file, then create the file.

```js
<%*
  const topic = await tp.system.prompt("Set topic", "this is a default value", true, false);
  const template = tp.file.find_tfile("Templates/Template2");
  await tp.file.create_new(template, topic);
-%>
```

The second template will whatever you want to actually be in the file you're creating.

---

Both of the above options are valid, depends on what you need.

**Comment by Deeds87** [comment]
Thanks for your help. Indeed i use `Templater: create new note from template`.

In the template i am working on:
1. The file will be created (Untitled.md)
2. I set some values like the projectCategory (suggester) and projectTitle (prompt).
3. No matter on which suggester i hit `esc` or cancel button the file will be deleted (project creation will interrupt)
4. **What i expected was the same behaviour when i cancel a prompt**

Here is the template i am working on:
``` js
<% "---" %>
<%*
/*  
================================================================================  
VERSION: 1.0

NAME: Neues_Projekt_TEMPLATE

GENERAL INFORMATION:
This template creates a new project with needed folders and a project dashboard.

REQUIRED:
The root folders of the vault have to follow these nameing syntax - xxx_FOLERNAME

DEPENDENCIES:
- Templater plugin
- getRootFolders.js script 
- trueOrFalseSuggester.js script 
- dataview plugin
- meta bind plugin
- minimal theme (cards layout)

TODO:

================================================================================
*/

//+++++++++++++++++++++++++++++      VARIABLES     +++++++++++++++++++++++++++++

// ----- Template config
const pathsToExclude = []; // Add paths to exclude from being a target location
/*
EXAMPLE:
let pathsToExclude = ["000_INBOX"] excludes 000_INBOX folder from being a
target location of the new project (used in getRootFolders script)
*/
let fileName = ""; // Dashboard file name
const rootFolders = await tp.user.getRootFolders(tp, pathsToExclude);
let categoryNames = [];
let categoryPaths = [];
const priorities = ["⭐", "⭐⭐", "⭐⭐⭐", "⭐⭐⭐⭐", "⭐⭐⭐⭐⭐"]

// ----- Metadata 
let projectCategory = ""; // Project category (vault root folders)
let projectTitle = ""; // Project name
let projectPriority = ""; // Project priority
let projectDueDate = ""; // Project due date
let projectMembers = "\n"; // People who are related to this project
let tags = []; // tags

//+++++++++++++++++++++++++++++      METADATA      +++++++++++++++++++++++++++++

// ----- PROJECT CATEGORY

// Keep attention to the required naming syntax of root folders (See info comment)
for(let folder of rootFolders){
	let folderName = folder.name.split("_")[1].toLowerCase();
	folderName = folderName.charAt(0).toUpperCase() + folderName.slice(1);
	categoryNames.push(folderName);
	categoryPaths.push(folder.path);
}

projectCategory = await tp.system.suggester(categoryNames, categoryNames, true,
	"Bitte wählen Sie eine Projektkategorie!");
// console.log("Project category: " + projectCategory);


// ----- PROJECT TITLE & FILE NAME

while(projectTitle == ""){
	projectTitle = await tp.system.prompt("Bitte einen Projekttitel eingeben",
		"", true, false);
	fileName = "Projektdashboard_" + projectTitle;
}

let isExistingProject = 
	await tp.file.exists(categoryPaths[categoryNames.indexOf(projectCategory)] + 
		"/Projekte/" + projectTitle + "/" + fileName + ".md");

while(isExistingProject){
	let change = 
		await tp.system.suggester(["Name", "Kategorie"],["changeName", "changeCategory"],
			true, "PROJEKT EXISTIERT BEREITS! - Name oder Kategorie ändern?");
		
	if(change == "changeName"){
		projectTitle = await tp.system.prompt("Neuen Projekttitel eingeben!",
			"", true, false);
		fileName = "Projekdashboard_" + projectTitle;
	} else if(change == "changeCategory"){
		projectCategory = await tp.system.suggester(categoryNames, categoryNames, 
			true, "Andere Kategorie wählen oder abbrechen!");
	}
	
	isExistingProject = 
		await tp.file.exists(categoryPaths[categoryNames.indexOf(projectCategory)] + 
			"/Projekte/" + projectTitle + "/" + fileName + ".md");
}
// console.log("Project title: " + projectTitle + " --- " + "File name: " + fileName)


// ----- PROJECT PRIORITY

while(projectPriority == ""){
	projectPriority = await tp.system.suggester(priorities, priorities, true,
		 "Bitte die Priorität des Projektes auswählen!");
}
//console.log("Project priority: " + projectPriority);


// ----- PROJECT DUE DATE 

let hasDueDate = 
	await tp.user.trueOrFalseSuggester(tp, "Möchten Sie ein Fälligkeitsdatum hinzufügen?");

projectDueDate = "Keine Fälligkeit";

if(hasDueDate){
	projectDueDate =
		await tp.system.prompt("Bitte geben Sie ein Fälligkeitsdatum ein! (DD.MM.YY)",
			"", true, false);
}
//console.log("Project due date: " + projectDueDate);


// ----- PROJECT MEMBERS

let hasMembers = 
	await tp.user.trueOrFalseSuggester(tp, "Möchten Sie Projektmitglieder hinzufügen?");

while(hasMembers){
	let newMember = 
		await tp.system.prompt("Projektmitglied hinzufügen!", "", true, false);
	projectMembers += `- "[[${newMember}]]"\n`;
	hasMembers = 
		await tp.user.trueOrFalseSuggester(tp, "Möchten Sie weitere Projektmitglieder hinzufügen?");
}
//console.log("Project members: " + projectMembers);


// ----- TAGS

tags.push(projectCategory, "Projekt");
let hasTags = 
	await tp.user.trueOrFalseSuggester(tp, "Möchten Sie Tags hinzufügen? (#{Kategorie}, #Projekt automatisch)");

while(hasTags){
	let newTag = await tp.system.prompt("Neues Tag hinzufügen!", "", true, false);
	tags.push(newTag);
	hasTags = 
		await tp.user.trueOrFalseSuggester(tp, "Möchten Sie weitere Tags hinzufügen?");
}
//console.log("Tags: " + tags);


//+++++++++++++++++++++++++ MOVE FILES & CREATE FOLDERS ++++++++++++++++++++++++
 
// ----- CREATE FOLDERS

let projectPath = categoryPaths[categoryNames.indexOf(projectCategory)] + 
	"/Projekte/" + projectTitle;
let projectFilesPath = projectPath + "/Projektdateien";
let projectAttachmentsPath = projectPath + "/Projektanhänge";

await this.app.vault.createFolder(projectFilesPath);
await this.app.vault.createFolder(projectAttachmentsPath);

// ----- MOVE FILE

await tp.file.move(projectPath + "/" + fileName);

%>

Kategorie: <% projectCategory %>
Status: Erstellt
Priorität: <% projectPriority %>
Fälligkeitsdatum: <% projectDueDate %>
Projektmitglieder: <% projectMembers %>
Erstellungsdatum: <% tp.file.creation_date("DD.MM.YY") %>
Zuletzt_Bearbeitet: <% tp.file.last_modified_date("DD.MM.YY - HH:mm") %>
tags: <% tags %>
cssclasses: list-cards, cards-cols-2

<% "---" %>

# <% projectTitle %>

- ##### Navigation
	- # [[Dashboard_Projekte]]

- ##### Tags 
	- `$= dv.current().file.tags`

- ##### Projektmitglieder
	- `$= dv.current().Projektmitglieder`

- ##### Projektinfo
	 - **Kategorie:**  `VIEW[{Kategorie}]`              
	 - **Priorität:**  `VIEW[{Priorität}]`   
	 - **Fälligkeitsdatum:**  `VIEW[{Fälligkeitsdatum}]`              
	 - **Status :**  `VIEW[{Status}]`    

---

```

**Comment by Zachatoo** [comment]
I didn't know cancelling a suggester would delete the file after it was created, that's cool!

Should be fixed in the latest release, 1.18.4.
